tensorflow_quantum/datasets/spin_system.py

```diff
@@ -0,0 +1,210 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Quantum datasets for several quantum many-body spin systems in the form of
+parameterized circuits"""
+import os
+import numpy as np
+import sympy
+import scipy.sparse as sps
+import cirq
+
+
+class SpinSystem(object):
+    """Abstract spin system class containing information about the quantum
+    data
+    """
+
+    def __init__(self, g: float, res_e: float, fidelity: float,
+                 gs_energy: float, parameters: np.ndarray, path: str):
+        """Instantiate this spin system.
+
+        Args:
+            g: Order parameter
+            res_e: Residual energy between the exact diagonalization ground
+            state and the circuit state.
+            fidelity: Fidelity between the exact diagonalization ground state
+            and the circuit state.
+            gs_energy: Exact diagonalization  round state energy
+            parameters: Numpy array of shape (M,depth) where M is the required
+             number of parameters per layer.
+            path: Path to the data.
+        """
+        self.g = g
+        self.res_e = res_e
+        self.fidelity = fidelity
+        self.gs_energy = gs_energy
+        self.parameters = parameters
+        self.path = path
+
+    def __lt__(self, other):
+        return self.g < other.g
+
+    def hamiltonian(self):
+        """Load the hamiltonian into a scipy sparse CSR matrix."""
+        return sps.load_npz(self.path + '/H.npz')
+
+    def ground_state(self):
+        """Load the ground state wave function into a complex numpy array."""
+        return np.load(self.path + '/groundstate.npy')
+
+
+def unique_name():
+    """
+    Generator to generate an infinite number of unique names.
+
+    Yields:
+        Strings of the form 'theta_<integer>'
+
+    """
+    num = 0
+    while True:
+        yield 'theta_' + str(num)
+        num += 1
+
+
+class SpinSystemDataset(object):
+    """Abstract class that handles a single instance of the system for a
+    single set of order parameters.
+    """
+
+    def __init__(self, nspins: int, system_name: str):
+        """Instantiate this data set.
+
+        Args:
+            nspins: The number of spins in the system.
+            system_name: Name of the system we want to load.
+        """
+
+        # TODO: How should downloading and storing the data be handled?
+        data_path = '/tmp/data/' + system_name + "_{}/".format(int(nspins))
```

Comment:
Prefer `tempfile.mkdtemp()`, it works on all platforms, "/tmp/" doesn't work on windows, for example.

Also note that temp directories are often cleared by reboots. Fo it you're expecting these to be big downloads you may want an option to set the save directory to somewhere durable, and/or a print out telling the iser where the data is.

Reply:
The /tmp/ folder was a temporary solution (pun intended). I rewrote everything using `pathlib`. The default directry is now at `~/tfq-datasets` and the user can specify their own directory with an argument.